### DISCUSSION TITLE
epochs number choice?

### DISCUSSION THREAD

**Discussion by En-J-A** [aub-mind/arabert#27] [Q&A] [answered]
How can I choose the optimal number for epochs in training? 
when I tried to put no of epoch = 10  15, or 20, I note during the loop, the max accuracy was at epoch = 2. 
but when I set no of epochs = 2, I can't get that high accuracy.

I note in your code examples the accuracy increases while epochs increases. 
but in my case it doesn't have one direction, like this output for epoch = 20 as an example 
```
Using deprecated `--per_gpu_train_batch_size` argument which will be removed in a future version. Using `--per_device_train_batch_size` is preferred.
Using deprecated `--per_gpu_train_batch_size` argument which will be removed in a future version. Using `--per_device_train_batch_size` is preferred.
Using deprecated `--per_gpu_eval_batch_size` argument which will be removed in a future version. Using `--per_device_eval_batch_size` is preferred.
Using deprecated `--per_gpu_train_batch_size` argument which will be removed in a future version. Using `--per_device_train_batch_size` is preferred.
Epoch: 100%
20/20 [48:27<00:00, 145.38s/it]
Iteration: 100%
864/864 [02:22<00:00, 6.07it/s]
{'loss': 0.24351061715020073, 'learning_rate': 1.9e-05, 'epoch': 1.0, 'step': 864}

/usr/local/lib/python3.6/dist-packages/torch/optim/lr_scheduler.py:200: UserWarning: Please also save or load the state of the optimzer when saving or loading the scheduler.
  warnings.warn(SAVE_STATE_WARNING, UserWarning)
Iteration: 100%
864/864 [02:24<00:00, 6.00it/s]
Using deprecated `--per_gpu_eval_batch_size` argument which will be removed in a future version. Using `--per_device_eval_batch_size` is preferred.
Evaluation: 100%
121/121 [00:52<00:00, 2.32it/s]

              precision    recall  f1-score   support

           0       0.82      0.55      0.66       640
           1       0.91      0.28      0.43       721
           2       0.44      0.98      0.61       574

    accuracy                           0.58      1935
   macro avg       0.72      0.60      0.57      1935
weighted avg       0.74      0.58      0.56      1935

[[352  20 268]
 [ 64 205 452]
 [ 11   1 562]]
{'eval_loss': 2.324988197601335, 'eval_f1_pos': 0.5661144694773962, 'eval_f1_neg': 0.5661144694773962, 'eval_f1_neu': 0.5661144694773962, 'eval_macro_f1': 0.5661144694773962, 'eval_macro_precision': 0.7232710510053528, 'eval_macro_recall': 0.6044737999391089, 'eval_accuracy': 0.5782945736434109, 'epoch': 1.1574074074074074, 'step': 1000}
{'loss': 0.1185383090266475, 'learning_rate': 1.8e-05, 'epoch': 2.0, 'step': 1728}

Iteration: 100%
864/864 [02:24<00:00, 5.98it/s]
Using deprecated `--per_gpu_eval_batch_size` argument which will be removed in a future version. Using `--per_device_eval_batch_size` is preferred.
Evaluation: 100%
121/121 [00:08<00:00, 14.76it/s]

              precision    recall  f1-score   support

           0       0.81      0.69      0.74       640
           1       0.84      0.64      0.73       721
           2       0.64      0.94      0.76       574

    accuracy                           0.74      1935
   macro avg       0.76      0.76      0.74      1935
weighted avg       0.77      0.74      0.74      1935

[[439  74 127]
 [ 79 459 183]
 [ 21  12 541]]
{'eval_loss': 1.3993108510283556, 'eval_f1_pos': 0.7430385421323368, 'eval_f1_neg': 0.7430385421323368, 'eval_f1_neu': 0.7430385421323368, 'eval_macro_f1': 0.7430385421323368, 'eval_macro_precision': 0.7641319190353323, 'eval_macro_recall': 0.7550206740581622, 'eval_accuracy': 0.7436692506459949, 'epoch': 2.314814814814815, 'step': 2000}
{'loss': 0.06103519157127098, 'learning_rate': 1.7e-05, 'epoch': 3.0, 'step': 2592}

Iteration: 100%
864/864 [02:24<00:00, 5.97it/s]
Using deprecated `--per_gpu_eval_batch_size` argument which will be removed in a future version. Using `--per_device_eval_batch_size` is preferred.
Evaluation: 100%
121/121 [00:04<00:00, 26.01it/s]

              precision    recall  f1-score   support

           0       0.81      0.59      0.68       640
           1       0.86      0.44      0.58       721
           2       0.50      0.97      0.66       574

    accuracy                           0.64      1935
   macro avg       0.72      0.66      0.64      1935
weighted avg       0.74      0.64      0.64      1935

[[378  45 217]
 [ 76 315 330]
 [ 14   5 555]]
{'eval_loss': 2.3934151222008038, 'eval_f1_pos': 0.6415707120009077, 'eval_f1_neg': 0.6415707120009077, 'eval_f1_neu': 0.6415707120009077, 'eval_macro_f1': 0.6415707120009077, 'eval_macro_precision': 0.7247785901292602, 'eval_macro_recall': 0.6648057195291094, 'eval_accuracy': 0.6449612403100775, 'epoch': 3.4722222222222223, 'step': 3000}
{'loss': 0.033778755753128616, 'learning_rate': 1.6000000000000003e-05, 'epoch': 4.0, 'step': 3456}

Iteration: 100%
864/864 [02:25<00:00, 5.96it/s]
Using deprecated `--per_gpu_eval_batch_size` argument which will be removed in a future version. Using `--per_device_eval_batch_size` is preferred.
Evaluation: 100%
121/121 [00:08<00:00, 14.28it/s]

              precision    recall  f1-score   support

           0       0.81      0.67      0.73       640
           1       0.86      0.55      0.67       721
           2       0.58      0.95      0.72       574

    accuracy                           0.71      1935
   macro avg       0.75      0.72      0.71      1935
weighted avg       0.76      0.71      0.71      1935

[[426  57 157]
 [ 84 399 238]
 [ 18   9 547]]
{'eval_loss': 2.030477235146408, 'eval_f1_pos': 0.7079792847941672, 'eval_f1_neg': 0.7079792847941672, 'eval_f1_neu': 0.7079792847941672, 'eval_macro_f1': 0.7079792847941672, 'eval_macro_precision': 0.7485207011557945, 'eval_macro_recall': 0.7239949102420983, 'eval_accuracy': 0.7090439276485788, 'epoch': 4.62962962962963, 'step': 4000}
{'loss': 0.019963370429144964, 'learning_rate': 1.5000000000000002e-05, 'epoch': 5.0, 'step': 4320}
.
.
.
.
.
.
.


              precision    recall  f1-score   support

           0       0.82      0.60      0.69       640
           1       0.84      0.54      0.65       721
           2       0.54      0.96      0.69       574

    accuracy                           0.68      1935
   macro avg       0.74      0.70      0.68      1935
weighted avg       0.75      0.68      0.68      1935

[[381  64 195]
 [ 67 386 268]
 [ 15   8 551]]
{'eval_loss': 3.0739450085743565, 'eval_f1_pos': 0.6798633372527613, 'eval_f1_neg': 0.6798633372527613, 'eval_f1_neu': 0.6798633372527613, 'eval_macro_f1': 0.6798633372527613, 'eval_macro_precision': 0.7363604777409388, 'eval_macro_recall': 0.6968701195550443, 'eval_accuracy': 0.6811369509043927, 'epoch': 18.51851851851852, 'step': 16000}
{'loss': 0.0008402223940248843, 'learning_rate': 1.0000000000000002e-06, 'epoch': 19.0, 'step': 16416}

Iteration: 100%
864/864 [02:29<00:00, 5.77it/s]
Using deprecated `--per_gpu_eval_batch_size` argument which will be removed in a future version. Using `--per_device_eval_batch_size` is preferred.
Evaluation: 100%
121/121 [00:04<00:00, 25.96it/s]

              precision    recall  f1-score   support

           0       0.83      0.52      0.64       640
           1       0.86      0.45      0.59       721
           2       0.48      0.98      0.65       574

    accuracy                           0.63      1935
   macro avg       0.72      0.65      0.62      1935
weighted avg       0.74      0.63      0.62      1935

[[333  48 259]
 [ 57 321 343]
 [  9   5 560]]
{'eval_loss': 3.6909896133547475, 'eval_f1_pos': 0.6241545408831674, 'eval_f1_neg': 0.6241545408831674, 'eval_f1_neu': 0.6241545408831674, 'eval_macro_f1': 0.6241545408831674, 'eval_macro_precision': 0.724934315687421, 'eval_macro_recall': 0.6470457450977075, 'eval_accuracy': 0.6273901808785529, 'epoch': 19.675925925925927, 'step': 17000}
{'loss': 0.0014848355893735533, 'learning_rate': 0.0, 'epoch': 20.0, 'step': 17280}


TrainOutput(global_step=17280, training_loss=0.028225022775155525)
```
and when I put no of epochs = 2 I got %68 as an accuracy

**Comment by WissamAntoun** [comment, marked as answer]
Usually it  takes betweet 2-6 epcohs to finetune. I suggest saving the model after each epoch by setting you save_steps parameter to 864. Then choose the model with the best validation score.

You might need to test the performance of the best model on an another test set

**Comment by En-J-A** [comment]
for testing the performance of the model I use the predict() function, do I need to apply the same steps on the dev dataset like the get_features function and passing dev_features to it?

**Comment by WissamAntoun** [comment]
yes for sure
